packages/realm/bindgen/src/templates/wrapper.ts

```diff
@@ -0,0 +1,405 @@
+        const hasInterestingDefault = ![undefined, "", "{}", "[]"].includes(field.defaultVal);
+
+        const docLines: string[] = [];
+        if (!field.isOptedInTo) {
+          docLines.push(
+            `@deprecated Add '${field.name}' to your opt-in list (under 'records/${record.name}/fields/') to use this.`,
+          );
+        }
+
+        if (hasInterestingDefault) {
+          docLines.push(`@default ${field.defaultVal}`);
+        }
+
+        let docComment = "";
+        if (docLines.length > 1) {
+          docComment = `/**\n * ${docLines.join("\n * ")}\n */\n`;
+        } else if (docLines.length === 1) {
+          docComment = `/** ${docLines[0]} */\n`;
+        }
+
+        return `${docComment} ${field.jsName}${optField ? "?" : ""}: ${generateType(spec, field.type, kind)};`;
+      });
+
+      return `export type ${record.jsName}${suffix(kind)} = {${fieldDeclarations.join("\n")}}`;
+    })
+    .join("\n");
+}
+
+function generateMethodDeclaration(spec: BoundSpec, method: Method) {
+  if (method instanceof Property) {
+    return `get ${method.jsName}(): ${generateType(spec, method.type, Kind.Return)}`;
+  } else {
+    const transformedSignature = method.sig.asyncTransformOrSelf();
+    const returnType = generateType(spec, transformedSignature.ret, Kind.Return);
+    const argumentDeclarations = transformedSignature.args.map(
+      (arg) => `${arg.name}: ${generateType(spec, arg.type, Kind.Argument)}`,
+    );
+    return (method.isStatic ? "static " : "") + method.jsName + `(${argumentDeclarations.join(", ")}): ${returnType};`;
+  }
+}
+
+function generateClassDeclaration(spec: BoundSpec, cls: Class) {
+  const methodDeclarations = cls.methods
+    .filter((method) => method.isOptedInTo)
+    .map(generateMethodDeclaration.bind(undefined, spec));
+  return `
+    export declare class ${cls.jsName} ${cls.base ? `extends ${cls.base.jsName}` : ""} {
+      /** Opting out of structural typing */
+      private brandFor${cls.jsName};
+      ${cls.subclasses.length === 0 ? "private" : "protected"} constructor();
+      ${methodDeclarations.join("\n")}
+      ${cls.iterable ? `[Symbol.iterator](): Iterator<${generateType(spec, cls.iterable, Kind.Return)}>;` : ""}
+    }`;
+}
+
+/**
+ * Generates code wrapping the native module, making it more suitable for consumption.
+ */
+export function generate({ spec: boundSpec, rawSpec, file }: TemplateContext): void {
+  // Check the support for primitives used
+  for (const primitive of rawSpec.primitives) {
+    if (primitive === "Mixed") continue;
+    assert(
+      Object.keys(PRIMITIVES_MAPPING).includes(primitive),
+      `Spec declares an unsupported primitive: "${primitive}"`,
+    );
+  }
+
+  // Check the support for template instances used
+  for (const template of Object.keys(rawSpec.templates)) {
+    assert(
+      Object.keys(TEMPLATE_MAPPING).includes(template),
+      `Spec declares an unsupported template instance: "${template}"`,
+    );
+  }
```

Comment:
Nit: Alternatively, this could be refactored into some `assertSupportedTypes()` (or the like) function (with the `rawSpec` param), and let `generate()` only focus on generating output.

Reply:
This is copy-paste from the former `typescript.ts` template: https://github.com/realm/realm-js/blob/ff3d889df4bb1da1ae40074cbcbea8bf8410223f/packages/realm/bindgen/src/templates/typescript.ts#L135-L150